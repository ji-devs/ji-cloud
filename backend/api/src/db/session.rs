use chrono::{DateTime, Utc};
use sqlx::PgConnection;
use tracing::instrument;
use uuid::Uuid;

use crate::token::SessionMask;

#[must_use]
fn generate_session_token() -> String {
    use rand::Rng;

    let mut bytes = [0_u8; 48];
    rand::thread_rng().fill(&mut bytes[..]);
    base64::encode_config(&bytes, base64::URL_SAFE)
}

#[instrument(skip(conn))]
pub async fn create(
    conn: &mut PgConnection,
    user_id: Uuid,
    valid_until: Option<&DateTime<Utc>>,
    mask: SessionMask,
    impersonator_id: Option<Uuid>,
) -> sqlx::Result<String> {
    let session = generate_session_token();
    sqlx::query!(
        "insert into session (token, user_id, impersonator_id, expires_at, scope_mask) values ($1, $2, $3, $4, $5)",
        &session,
        user_id,
        impersonator_id,
        valid_until,
        mask.bits()
    )
    .execute(conn).await?;

    Ok(session)
}

#[instrument(skip(conn))]
pub async fn clear_any(
    conn: &mut PgConnection,
    user_id: Uuid,
    mask: SessionMask,
) -> sqlx::Result<()> {
    sqlx::query!(
        "delete from session where user_id = $1 and (scope_mask | $2) <> 0",
        user_id,
        mask.bits(),
    )
    .execute(conn)
    .await?;

    Ok(())
}

/// finds a one time session and deletes it after verifying its valididity.
#[instrument(skip_all)]
pub async fn get_onetime(
    txn: &mut PgConnection,
    min_mask: SessionMask,
    token: &str,
) -> sqlx::Result<Option<Uuid>> {
    let res = sqlx::query!(
        r#"delete from session where token = $1 and (scope_mask & $2) = $2 returning user_id"#,
        token,
        min_mask.bits()
    )
    .fetch_optional(txn)
    .await?;

    Ok(res.map(|it| it.user_id))
}

#[instrument(skip_all)]
pub async fn delete(txn: &mut PgConnection, token: &str) -> sqlx::Result<()> {
    sqlx::query!("delete from session where token = $1", &token)
        .execute(txn)
        .await?;

    Ok(())
}
