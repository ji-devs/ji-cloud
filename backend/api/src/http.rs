use std::{net::TcpListener, sync::Arc};

use actix_service::Service;
use actix_web::{
    dev::{MessageBody, Server, ServiceRequest, ServiceResponse},
    error::Error,
    middleware::{Compat, Condition},
    web::{method, Data},
    HttpResponse,
};
use core::{
    config::JSON_BODY_LIMIT,
    env::env_bool,
    http::{get_addr, get_tcp_fd},
    settings::RuntimeSettings,
};
use futures::Future;
use sqlx::postgres::PgPool;
use tracing::Span;
use tracing_actix_web::{root_span, DefaultRootSpanBuilder, RootSpanBuilder, TracingLogger};

use crate::{
    error::BasicError,
    service::{self, mail, s3, upload::cleaner, ServiceData},
    translate,
};

mod cors;
mod endpoints;

fn log_ise<B: MessageBody, T>(
    request: ServiceRequest,
    srv: &T,
) -> impl Future<Output = actix_web::Result<T::Response>>
where
    T: Service<ServiceRequest, Response = ServiceResponse<B>, Error = actix_web::Error>,
{
    let uri: serde_json::Value = request.uri().to_string().into();
    let method: serde_json::Value = request.method().to_string().into();

    let future = srv.call(request);
    async {
        let mut result = future.await?;
        if result.status() == 500 {
            let response: &mut actix_web::HttpResponse<_> = result.response_mut();

            if let Some(err) = response.extensions_mut().remove::<anyhow::Error>() {
                log::error!("ISE while responding to request: {:?}", err);
                sentry::add_breadcrumb(sentry::Breadcrumb {
                    ty: "http".to_owned(),
                    category: Some("request".into()),
                    data: {
                        let mut map = sentry::protocol::Map::new();
                        map.insert("url".to_owned(), uri);
                        map.insert("method".to_owned(), method);
                        map
                    },
                    ..Default::default()
                });

                sentry::integrations::anyhow::capture_anyhow(&err);
            }
        }

        Ok(result)
    }
}

pub struct Application {
    port: u16,
    server: Option<Server>,
}

impl Application {
    pub fn new(port: u16, server: Server) -> Self {
        Self {
            port,
            server: Some(server),
        }
    }

    pub fn port(&self) -> u16 {
        self.port
    }

    // A more expressive name that makes it clear that
    // this function only returns when the application is stopped.
    pub async fn run_until_stopped(mut self) -> Result<(), std::io::Error> {
        if let Some(server) = self.server.take() {
            server.await?;
        }

        Ok(())
    }

    pub async fn stop(mut self, graceful: bool) {
        if let Some(server) = self.server.take() {
            server.stop(graceful).await
        }
    }
}

impl Drop for Application {
    fn drop(&mut self) {
        if let Some(server) = self.server.take() {
            let _ = tokio::spawn(server.stop(false));
        }
    }
}

#[actix_web::main]
pub async fn build_and_run(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<()> {
    let app = build(
        pool,
        settings,
        s3,
        gcp_key_store,
        gcs,
        algolia,
        algolia_key_store,
        jwk_verifier,
        mail_client,
        algolia_manager,
        media_upload_cleaner,
        google_translate,
    )?;
    app.run_until_stopped().await?;

    Ok(())
}

pub fn build(
    pool: PgPool,
    settings: RuntimeSettings,
    s3: Option<s3::Client>,
    gcp_key_store: Option<service::GcpAccessKeyStore>,
    gcs: Option<service::storage::Client>,
    algolia: Option<crate::algolia::Client>,
    algolia_key_store: Option<crate::algolia::SearchKeyStore>,
    jwk_verifier: Arc<crate::jwk::JwkVerifier>,
    mail_client: Option<mail::Client>,
    algolia_manager: Option<crate::algolia::Manager>,
    media_upload_cleaner: Option<cleaner::UploadCleaner>,
    google_translate: Option<translate::GoogleTranslate>,
) -> anyhow::Result<Application> {
    let local_insecure = settings.is_local();
    let api_port = settings.api_port;

    let s3 = s3.map(ServiceData::new);
    let gcp_key_store = gcp_key_store.map(ServiceData::new);
    let gcs = gcs.map(ServiceData::new);
    let algolia = algolia.map(ServiceData::new);
    let algolia_key_store = algolia_key_store.map(ServiceData::new);
    let mail_client = mail_client.map(ServiceData::new);
    let algolia_manager = algolia_manager.map(ServiceData::new);
    let media_upload_cleaner = media_upload_cleaner.map(ServiceData::new);
    let google_translate = google_translate.map(ServiceData::new);

    let server = actix_web::HttpServer::new(move || {
        let app = actix_web::App::new()
            .app_data(Data::new(pool.clone()))
            .app_data(Data::new(settings.clone()));

        let app = match s3.clone() {
            Some(s3) => app.app_data(s3),
            None => app,
        };

        let app = match gcp_key_store.clone() {
            Some(gcp_key_store) => app.app_data(gcp_key_store),
            None => app,
        };

        let app = match gcs.clone() {
            Some(gcs) => app.app_data(gcs),
            None => app,
        };

        let app = match algolia.clone() {
            Some(algolia) => app.app_data(algolia),
            None => app,
        };

        let app = match algolia_key_store.clone() {
            Some(algolia_key_store) => app.app_data(algolia_key_store),
            None => app,
        };

        let app = match mail_client.clone() {
            Some(mail_client) => app.app_data(mail_client),
            None => app,
        };

        let app = match algolia_manager.clone() {
            Some(algolia_manager) => app.app_data(algolia_manager),
            None => app,
        };

        let app = match media_upload_cleaner.clone() {
            Some(media_upload_cleaner) => app.app_data(media_upload_cleaner),
            None => app,
        };
        let app = match google_translate.clone() {
            Some(google_translate) => app.app_data(google_translate),
            None => app,
        };

        let enable_tracing_logs = env_bool("ENABLE_TRACING_LOGS");

        app.app_data(Data::from(jwk_verifier.clone()))
            .wrap(cors::get(local_insecure))
            .wrap(Condition::new(
                !enable_tracing_logs,
                Compat::new(actix_web::middleware::Logger::default()),
            ))
            .wrap(Condition::new(
                enable_tracing_logs,
                Compat::new(TracingLogger::<JigziSpanBuilder>::new()),
            ))
            .wrap_fn(log_ise)
            .app_data(
                actix_web::web::JsonConfig::default()
                    .limit(JSON_BODY_LIMIT as usize)
                    .error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::QueryConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .app_data(
                actix_web::web::PathConfig::default().error_handler(|_, _| bad_request_handler()),
            )
            .external_resource(
                "google_cloud_oauth",
                "https://accounts.google.com/o/oauth2/v2/auth",
            )
            .default_service(actix_web::web::to(default_route))
            .configure(endpoints::user::configure)
            .configure(endpoints::category::configure)
            .configure(endpoints::image::configure)
            .configure(endpoints::audio::configure)
            .configure(endpoints::meta::configure)
            .configure(endpoints::jig::configure)
            .configure(endpoints::module::configure)
            .configure(endpoints::additional_resource::configure)
            .configure(endpoints::jig::report::configure)
            .configure(endpoints::jig::curation::configure)
            .configure(endpoints::course::configure)
            .configure(endpoints::admin::configure)
            .configure(endpoints::animation::configure)
            .configure(endpoints::search::configure)
            .configure(endpoints::media::configure)
            .configure(endpoints::session::configure)
            .configure(endpoints::locale::configure)
            .configure(endpoints::scheduler::configure)
            .configure(endpoints::pdf::configure)
            .configure(endpoints::badge::configure)
            .route("/", method(http::Method::GET).to(no_content_response))
    });

    // if listenfd doesn't take a TcpListener (i.e. we're not running via
    // the command above), we fall back to explicitly binding to a given
    // host:port.
    let listener = if let Some(l) = get_tcp_fd() {
        l
    } else {
        TcpListener::bind(get_addr(Some(api_port)))?
    };

    let port = listener.local_addr().unwrap().port();

    let server = server.listen(listener)?;

    Ok(Application::new(port, server.run()))
}

fn default_route() -> HttpResponse {
    HttpResponse::NotFound().json(BasicError::with_message(
        http::StatusCode::NOT_FOUND,
        "Route not found".to_owned(),
    ))
}

async fn no_content_response() -> HttpResponse {
    HttpResponse::NoContent().finish()
}

pub fn bad_request_handler() -> actix_web::Error {
    BasicError::new(http::StatusCode::BAD_REQUEST).into()
}

pub struct JigziSpanBuilder;

impl RootSpanBuilder for JigziSpanBuilder {
    fn on_request_start(request: &ServiceRequest) -> Span {
        let route = request.match_pattern().unwrap_or("default".to_string());
        let transaction_name = format!("{} {}", request.method(), route);
        let transaction_name = transaction_name.as_str();
        root_span!(request, message = transaction_name)
    }

    fn on_request_end<B>(span: Span, outcome: &Result<ServiceResponse<B>, Error>) {
        DefaultRootSpanBuilder::on_request_end(span, outcome)
    }
}
