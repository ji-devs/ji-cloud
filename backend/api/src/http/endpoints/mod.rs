pub mod additional_resource;
pub mod admin;
pub mod animation;
pub mod audio;
pub mod category;
pub mod image;
pub mod jig;
pub mod locale;
pub mod media;
pub mod meta;
pub mod module;
pub mod search;
pub mod session;
pub mod user;
