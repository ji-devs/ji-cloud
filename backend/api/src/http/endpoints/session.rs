use actix_web::{
    web::{Data, ServiceConfig},
    HttpResponse,
};
use chrono::{Duration, Utc};
use core::settings::RuntimeSettings;
use shared::{
    api::{endpoints::session, ApiEndpoint, PathParts},
    domain::session::{CreateSessionResponse, NewSessionResponse},
};
use sqlx::PgPool;

use crate::{
    db,
    domain::{NoContentClearAuth, RegistrationStatus},
    error,
    extractor::{EmailBasicUser, SessionAny, TokenSessionOf},
    token::{create_auth_token, SessionMask},
};

mod oauth;

/// Login with basic authorization.
/// May return resources for *signing up* if the user doesn't have a profile.
async fn create_session(
    db: Data<PgPool>,
    settings: Data<RuntimeSettings>,
    user: EmailBasicUser,
) -> Result<HttpResponse, error::Server> {
    let login_ttl = settings
        .login_token_valid_duration
        .unwrap_or(Duration::weeks(2));

    let (mask, valid_until) = match user.registration_status {
        RegistrationStatus::New => panic!("This isn't currently possible"),
        RegistrationStatus::Validated => (
            SessionMask::PUT_PROFILE | SessionMask::DELETE_ACCOUNT,
            Utc::now() + Duration::hours(1),
        ),
        RegistrationStatus::Complete => (SessionMask::GENERAL, Utc::now() + login_ttl),
    };

    let mut txn = db.begin().await?;

    let session = db::session::create(&mut txn, user.id, Some(&valid_until), mask, None).await?;

    let (csrf, cookie) = create_auth_token(
        &settings.token_secret,
        settings.is_local(),
        login_ttl,
        &session,
    )?;

    txn.commit().await?;

    let response = NewSessionResponse { csrf };

    let response = if !mask.contains(SessionMask::GENERAL) {
        CreateSessionResponse::Register {
            response,
            oauth_profile: None,
        }
    } else {
        CreateSessionResponse::Login(response)
    };

    Ok(HttpResponse::Created().cookie(cookie).json(response))
}

/// Logout
async fn delete_session(
    db: Data<PgPool>,
    session: TokenSessionOf<SessionAny>,
) -> Result<NoContentClearAuth, error::Server> {
    sqlx::query!("delete from session where token = $1", session.claims.token)
        .execute(db.as_ref())
        .await?;

    Ok(NoContentClearAuth)
}

pub fn configure(cfg: &mut ServiceConfig) {
    cfg.route(
        <session::GetOAuthUrl as ApiEndpoint>::Path::PATH,
        session::GetOAuthUrl::METHOD.route().to(oauth::get_url),
    )
    .route(
        <session::Create as ApiEndpoint>::Path::PATH,
        session::Create::METHOD.route().to(create_session),
    )
    .route(
        <session::Delete as ApiEndpoint>::Path::PATH,
        session::Delete::METHOD.route().to(delete_session),
    )
    .route(
        <session::CreateOAuth as ApiEndpoint>::Path::PATH,
        session::CreateOAuth::METHOD.route().to(oauth::create),
    );
}
