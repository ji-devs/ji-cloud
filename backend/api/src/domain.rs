use actix_http::body::BoxBody;
use actix_web::{
    cookie::{Cookie, SameSite},
    HttpResponse, Responder,
};

use crate::{
    db::{self, account::AccountMember},
    error,
};
use chrono::{DateTime, Utc};
use http::StatusCode;
use shared::domain::billing::AccountId;
use shared::domain::user::{UserId, UserScope};
use shared::domain::{
    category::{Category, CategoryId},
    session::AUTH_COOKIE_NAME,
};
use sqlx::PgPool;
use std::{cell::RefCell, convert::TryFrom, rc::Rc};
use std::{collections::HashMap, fmt};
use tracing::instrument;
use uuid::Uuid;

#[derive(Debug)]
pub struct RawCategory {
    pub id: Uuid,
    pub parent_id: Option<Uuid>,
    pub name: String,
    pub index: i16,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_scopes: Vec<i16>,
}

#[derive(Debug, Clone)]
struct CategoryNode {
    pub id: Uuid,
    pub name: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_scopes: Vec<i16>,
    pub children: Vec<Rc<RefCell<CategoryNode>>>,
}

impl From<CategoryNode> for Category {
    fn from(category_node: CategoryNode) -> Self {
        Self {
            id: CategoryId(category_node.id),
            name: category_node.name,
            created_at: category_node.created_at,
            updated_at: category_node.updated_at,
            user_scopes: {
                category_node
                    .user_scopes
                    .into_iter()
                    .map(|x| UserScope::try_from(x).expect("detected an invalid user scope"))
                    .collect()
            },
            children: {
                category_node
                    .children
                    .iter()
                    .map(|category| {
                        let category: CategoryNode = (**category).borrow().clone();
                        category.into()
                    })
                    .collect()
            },
        }
    }
}

#[instrument(skip_all)]
pub fn build_tree(categories: Vec<RawCategory>) -> Vec<Category> {
    let mut nodes: Vec<Rc<RefCell<CategoryNode>>> = Vec::new();
    let mut lookup: HashMap<Uuid, Rc<RefCell<CategoryNode>>> = HashMap::new();

    // Now we know this category exists in the lookup table
    for raw in categories.iter() {
        lookup.insert(
            raw.id.clone(),
            Rc::new(RefCell::new(CategoryNode {
                id: raw.id.clone(),
                name: raw.name.clone(),
                created_at: raw.created_at,
                updated_at: raw.updated_at,
                user_scopes: raw.user_scopes.clone(),
                children: Vec::new(),
            })),
        );
    }

    for raw in categories.iter() {
        let current_node = lookup
            .get(&raw.id)
            .expect("Category wasn't added to lookup map");
        match raw.parent_id {
            None => {
                // This is a root category, add it to the nodes list
                nodes.push(current_node.clone());
            }
            Some(parent_id) => {
                let parent_node = lookup
                    .get(&parent_id)
                    .expect("Parent caregory wasn't added to lookup map");
                let mut parent_node = (**parent_node).borrow_mut();
                parent_node.children.push(current_node.clone());
            }
        }
    }

    nodes
        .iter()
        .map(|node| {
            let category: CategoryNode = (**node).borrow().clone();
            let category: Category = category.into();
            category
        })
        .collect()
}

#[derive(Debug, Copy, Clone, sqlx::Type)]
#[repr(u16)]
pub enum RegistrationStatus {
    /// The user was just registered ([`POST /v1/user`](shared::api::endpoints::user::Register))
    ///
    /// This state is *skipped* via ouath (they're automatically [`Validated`](Self::Validated))
    /// In this state the user is allowed to:
    /// * Request a new verification email
    /// * Request a password reset(?)
    New = 0,

    /// The user has gone through OAuth and created a new user, or has finished email verification.
    ///
    /// In this state the user is allowed to:
    /// * Request a password reset
    /// * Complete registration via creating their profile
    /// * Delete their account
    Validated = 1,

    /// The user has *completely* finished signup, and their profile has been completed.
    ///
    /// In this state the user is allowed to do anything that their scopes allow.
    /// All users can:
    /// * Change their email (not implemented nor currently on the roadmap)
    /// * Change their password if logged in, reset their password if not.
    /// * Access the API
    /// * Change their profile details (not currently implemented)
    /// * Delete their account
    Complete = 2,
}

#[derive(Debug)]
pub struct NoContentClearAuth;

impl fmt::Display for NoContentClearAuth {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("No Content")
    }
}

impl Responder for NoContentClearAuth {
    type Body = BoxBody;

    fn respond_to(self, _: &actix_web::HttpRequest) -> HttpResponse<BoxBody> {
        let mut cookie = Cookie::named(AUTH_COOKIE_NAME);
        cookie.set_max_age(time::Duration::seconds(0));
        cookie.set_http_only(true);
        cookie.set_same_site(SameSite::Strict);

        actix_web::HttpResponse::build(StatusCode::NO_CONTENT)
            .content_type("application/octet-stream")
            .cookie(cookie)
            .finish()
    }
}

#[derive(Debug)]
pub enum UserAuthorization {
    SystemAdministrator,
    AccountAdministrator,
    AccountMember,
}

impl UserAuthorization {
    pub fn is_system_administrator(&self) -> bool {
        matches!(self, UserAuthorization::SystemAdministrator)
    }

    pub fn test_authorized(&self, require_account_admin: bool) -> Result<(), error::Account> {
        if self.is_authorized(require_account_admin) {
            Ok(())
        } else {
            Err(error::Account::Forbidden)
        }
    }

    pub fn is_authorized(&self, require_account_admin: bool) -> bool {
        match self {
            UserAuthorization::AccountMember if require_account_admin => false,
            _ => true,
        }
    }
}

pub async fn user_authorization(
    db: &PgPool,
    user_id: &UserId,
    account_id: &AccountId,
) -> Result<UserAuthorization, error::Account> {
    Ok(
        if db::user::has_scopes(db, *user_id, &[UserScope::Admin]).await? {
            UserAuthorization::SystemAdministrator
        } else {
            match db::account::user_account_membership(db, user_id, account_id).await? {
                Some(AccountMember::Admin) => UserAuthorization::AccountAdministrator,
                Some(AccountMember::User) => UserAuthorization::AccountMember,
                None => return Err(error::Account::Forbidden),
            }
        },
    )
}
