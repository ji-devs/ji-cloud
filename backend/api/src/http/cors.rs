use actix_web::http::{header, Method};
use core::config::CORS_ORIGINS;

pub fn get(local_insecure: bool) -> actix_cors::Cors {
    let mut cors = actix_cors::Cors::default()
        .supports_credentials()
        .allowed_methods(&[
            Method::GET,
            Method::POST,
            Method::PATCH,
            Method::DELETE,
            Method::PUT,
            Method::OPTIONS,
        ])
        .allowed_headers(&[
            header::AUTHORIZATION,
            header::CONTENT_TYPE,
            header::HeaderName::from_static("x-csrf"),
        ]);

    if local_insecure {
        cors = cors.allow_any_origin();
    } else {
        for origin in CORS_ORIGINS {
            cors = cors.allowed_origin(origin);
        }
    }

    cors
}
