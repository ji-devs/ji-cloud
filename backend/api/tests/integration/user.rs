use crate::{
    fixture::Fixture,
    helpers::{initialize_server, initialize_server_and_get_db, LoginExt},
    service::{self, Service},
};
use chrono::{Duration, Utc};
use http::StatusCode;
use ji_cloud_api::http::Application;
use serde_json::json;
use shared::domain::session::NewSessionResponse;
use shared::domain::{
    meta::{AffiliationId, AgeRangeId, SubjectId},
    session::CreateSessionResponse,
    user::PutProfileRequest,
};
use sqlx::PgPool;

mod color;
mod font;

#[actix_rt::test]
async fn get_profile() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/user/me/profile", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body);

    Ok(())
}

#[actix_rt::test]
async fn put_profile() -> anyhow::Result<()> {
    if !service::email_test_guard() {
        return Ok(());
    }

    const SUB: &str = "Sa84_qiKlh7WbOxeR9lofYJngysK_unF";
    let csrf: &str = "FOYKzUtD7wCLb7JJ";
    let key = &**super::helpers::PASETO_KEY;

    // Generate auth token that the server will accept.
    // On the front-end this below is handled by basic auth through `POST /v1/session`, which
    // calls this same function on the backend and returns the csrf + cookie used to authenticate.
    // See `basic_auth_flow`
    let token = ji_cloud_api::token::create_auth_token_no_cookie(
        key,
        Duration::minutes(10),
        SUB,
        csrf.to_owned(),
        Utc::now(),
    )
    .expect("failed to create auth token");

    // test server application
    let app = initialize_server(&[Fixture::User], &[Service::Email]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    // put profile
    let resp = client
        .put(&format!("http://0.0.0.0:{}/v1/user/me/profile", port))
        .header("X-CSRF", csrf)
        .header("Cookie", format!("X-AUTH={}", token))
        .json(&PutProfileRequest {
            username: "test_user".to_owned(),
            over_18: true,
            given_name: "name".to_owned(),
            family_name: "nameson".to_owned(),
            language: "en_US".to_owned(),
            locale: "en_US".to_owned(),
            timezone: chrono_tz::America::Los_Angeles,
            opt_into_edu_resources: true,
            organization: None,
            subjects: Vec::<SubjectId>::new(),
            age_ranges: Vec::<AgeRangeId>::new(),
            affiliations: Vec::<AffiliationId>::new(),
            location: None,
        })
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".csrf" => "[csrf]"});

    Ok(())
}

#[ignore]
#[actix_rt::test]
async fn verify_email() -> anyhow::Result<()> {
    if !service::email_test_guard() {
        return Ok(());
    }

    let app = initialize_server(&[Fixture::User], &[Service::Email]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/user/verify-email", port))
        .json(&json!({
            "verify": {
                "token": "L6gfXvgZeUBt8pdmLBnsGPEWUe3qGCK2_DF"
            }
        }))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    Ok(())
}

#[actix_rt::test]
async fn basic_auth_flow_no_login() -> anyhow::Result<()> {
    if !service::email_test_guard() {
        return Ok(());
    }

    const EMAIL: &str = "testfakeemailnotreal@iiiiiis.test";
    const PASSWORD: &str = "badpassword";

    let (app, db): (Application, PgPool) =
        initialize_server_and_get_db(&[], &[Service::Email]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    // 0. register basic auth user
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/user", port))
        .json(&json!({
            "email": EMAIL,
            "password": PASSWORD,
        }))
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    // 1. request resend email
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/user/verify-email", port))
        .json(&json!({
            "resend": {"email": "testfakeemailnotreal@iiiiiis.test" }
        }))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    // 1.1. retrieve verification token through database directly
    let mut txn = db.begin().await?;
    let token = sqlx::query!(r#"select token from "session" order by created_at limit 1"#)
        .fetch_one(&mut txn)
        .await?
        .token;

    // 2. verify email
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/user/verify-email", port))
        .json(&json!({
            "verify": { "token": token }
        }))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    // 3. Create user profile for the first time

    // 3.1. extract auth info from response
    let token = resp.cookies().next().unwrap().value().to_owned();
    log::info!("{:?}", &token);
    let body = resp.json::<NewSessionResponse>().await?;
    log::info!("{:?}", &body);

    // 3.2. put profile
    let resp = client
        .put(&format!("http://0.0.0.0:{}/v1/user/me/profile", port))
        .header("X-CSRF", body.csrf.as_str())
        .header("Cookie", format!("X-AUTH={}", token))
        .json(&PutProfileRequest {
            username: "test_user".to_owned(),
            over_18: true,
            given_name: "name".to_owned(),
            family_name: "nameson".to_owned(),
            language: "en_US".to_owned(),
            locale: "en_US".to_owned(),
            timezone: chrono_tz::America::Los_Angeles,
            opt_into_edu_resources: true,
            organization: None,
            subjects: Vec::<SubjectId>::new(),
            age_ranges: Vec::<AgeRangeId>::new(),
            affiliations: Vec::<AffiliationId>::new(),
            location: None,
        })
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;
    txn.commit().await?;

    insta::assert_json_snapshot!(body, {".csrf" => "[csrf]"});

    Ok(())
}

#[actix_rt::test]
async fn basic_auth_flow() -> anyhow::Result<()> {
    if !service::email_test_guard() {
        return Ok(());
    }

    const EMAIL: &str = "testfakeemailnotreal@iiiiiis.test";
    const PASSWORD: &str = "badpassword";

    let (app, db): (Application, PgPool) =
        initialize_server_and_get_db(&[], &[Service::Email]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    // 0. register basic auth user
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/user", port))
        .json(&json!({
            "email": EMAIL,
            "password": PASSWORD,
        }))
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    // 1. request resend email
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/user/verify-email", port))
        .json(&json!({
            "resend": {"email": "testfakeemailnotreal@iiiiiis.test" }
        }))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    // 1.1. retrieve verification token through database directly
    let mut txn = db.begin().await?;
    let token = sqlx::query!(r#"select token from "session" order by created_at limit 1"#)
        .fetch_one(&mut txn)
        .await?
        .token;

    // 2. verify email
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/user/verify-email", port))
        .json(&json!({
            "verify": { "token": token }
        }))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    // 3. login via basic auth `POST /v1/session`

    // 3.1. bad password/username. rejected login
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/session", port))
        .basic_auth(EMAIL, Some("aasdasda"))
        .send()
        .await?;

    assert_eq!(resp.status(), StatusCode::UNAUTHORIZED);

    // 3.2. basic auth login successful
    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/session", port))
        .basic_auth(EMAIL, Some(PASSWORD))
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    // 4. Create user profile for the first time

    // 4.1. extract auth info from response
    let token = resp.cookies().next().unwrap().value().to_owned();
    let body = resp.json::<CreateSessionResponse>().await?;

    let csrf = match body {
        CreateSessionResponse::Register(resp) => resp.csrf,
        _ => {
            return Err(anyhow::anyhow!(
                "invalid session response to create a user profile!"
            ))
        }
    };

    // 4.2. put profile
    let resp = client
        .put(&format!("http://0.0.0.0:{}/v1/user/me/profile", port))
        .header("X-CSRF", csrf.as_str())
        .header("Cookie", format!("X-AUTH={}", token))
        .json(&PutProfileRequest {
            username: "test_user".to_owned(),
            over_18: true,
            given_name: "name".to_owned(),
            family_name: "nameson".to_owned(),
            language: "en_US".to_owned(),
            locale: "en_US".to_owned(),
            timezone: chrono_tz::America::Los_Angeles,
            opt_into_edu_resources: true,
            organization: None,
            subjects: Vec::<SubjectId>::new(),
            age_ranges: Vec::<AgeRangeId>::new(),
            affiliations: Vec::<AffiliationId>::new(),
            location: None,
        })
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;
    txn.commit().await?;

    insta::assert_json_snapshot!(body, {".csrf" => "[csrf]"});

    Ok(())
}
