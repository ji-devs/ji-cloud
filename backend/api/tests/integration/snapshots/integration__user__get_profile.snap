---
source: tests/integration/user.rs
expression: body

---
{
  "id": "1f241e1b-b537-493f-a230-075cb16315be",
  "username": "test",
  "email": "test@test.test",
  "given_name": "Bobby",
  "family_name": "Tables",
  "profile_image_id": null,
  "language": "en_US",
  "locale": "en_US",
  "opt_into_edu_resources": true,
  "over_18": true,
  "timezone": "US/Pacific-New",
  "scopes": [
    "Admin"
  ],
  "created_at": "2020-08-08T00:11:21Z",
  "organization": "test org"
}
